### PR TITLE
Add Options parameter to soft delete functions (#187)

### PR DESCRIPTION
**Pull request by danicodes01**
- Adds optional opts parameter to soft_delete/2, soft_deleteand soft_delete_all/2
- Added tests to cover the new functionality
- Maintaions backward compatibility with existing code
- Enables use of options like 'prefix' for multi-tenant databases
- Updates function documentation with examples showing options usage
- Adds section to README about using options with soft delete functions

### CODE REVIEW

File: test/soft_delete_repo_test.exs [hunk 1]
```diff
@@ -100,6 +124,24 @@ defmodule Ecto.SoftDelete.Repo.Test do
...
+
+      assert User |> Repo.all(with_deleted: true) |> length() == 2
+
+      assert %DateTime{} =
+               Repo.get_by!(User, [email: "test0@example.com"], with_deleted: true).deleted_at
+
+      assert %DateTime{} =
+               Repo.get_by!(User, [email: "test1@example.com"], with_deleted: true).deleted_at
+    end
+  end
```
Comment (stuartjohnpage) [inline test/soft_delete_repo_test.exs:156]: My only thought would be: it could be good to test with an actual Ecto option, and make sure it gets passed through. I know the point was to pass a specific Ecto option (`:prefix`), but we could try something else, eg:

```
    test "soft deletes with log option" do
      Repo.insert!(%User{email: "test0@example.com"})

      log = ExUnit.CaptureLog.capture_log(fn ->
        Repo.soft_delete_all(User, log: :info)
      end)

      assert log =~ "UPDATE"
      assert log =~ "deleted_at"
    end
    ```

    but it's probably not the end of the world either way!

File: test/soft_delete_repo_test.exs [hunk 2]
```diff
@@ -75,6 +87,18 @@ defmodule Ecto.SoftDelete.Repo.Test do
...
+      user = Repo.insert!(%User{email: "test0@example.com"})
+
+      # Test with empty options list
+      assert {:ok, %User{}} = Repo.soft_delete(user, [])
+
+      assert %DateTime{} =
+               Repo.get_by!(User, [email: "test0@example.com"], with_deleted: true).deleted_at
+    end
+  end
+
```
Comment (stuartjohnpage) [inline test/soft_delete_repo_test.exs:102]: same here, see below!

File: lib/ecto/soft_delete_repo.ex [hunk 3]
```diff
@@ -41,35 +53,63 @@ defmodule Ecto.SoftDelete.Repo do
...
-      def soft_delete(struct_or_changeset) do
-        struct_or_changeset
-        |> Ecto.Changeset.change(deleted_at: DateTime.utc_now())
-        |> update()
+      # Define soft_delete and soft_delete! with options
+      def soft_delete(struct_or_changeset, opts \\ []) do
+        changeset = Ecto.Changeset.change(struct_or_changeset, deleted_at: DateTime.utc_now())
+
+        case opts do
+          # Original behavior without options
```
Comment (stuartjohnpage) [inline lib/ecto/soft_delete_repo.ex]: you can probably nix this comment and the one below

**Review by stuartjohnpage — APPROVED** [review]
I think this looks great: a nit on the comments, and maybe adding a test case where the KW list isn't just empty, but otherwise great! Thank you

File: lib/ecto/soft_delete_repo.ex [hunk 4]
```diff
@@ -41,35 +53,63 @@ defmodule Ecto.SoftDelete.Repo do
...
+      def soft_delete!(struct_or_changeset, opts \\ []) do
+        changeset = Ecto.Changeset.change(struct_or_changeset, deleted_at: DateTime.utc_now())
+
+        case opts do
+          [] ->
+            changesetForUpdate = changeset
+            __MODULE__.update!(changesetForUpdate)
+          _ ->
+            __MODULE__.update!(changeset, opts)
+        end
```
Comment (stuartjohnpage) [inline lib/ecto/soft_delete_repo.ex]: I think you can simplify this a bit too: there's also some sneaky little javascript syntax and an unnecessary var that can be ⚡ :
```
  case opts do
    [] -> __MODULE__.update!(changeset)
    _ -> __MODULE__.update!(changeset, opts)
  end
  ```
  Credit to @gk-per

**Review by stuartjohnpage — DISMISSED** [review]

File: lib/ecto/soft_delete_repo.ex (hunk 4 above)
Comment (danicodes01) [reply lib/ecto/soft_delete_repo.ex]: Got it. Working on the fix now to better follow Elixir style conventions and make the function more readable. Thank you!

**Review by stuartjohnpage — APPROVED** [review]
This looks good to me! Thank you for your contribution and speedy responses

File: lib/ecto/soft_delete_repo.ex [hunk 5]
```diff
@@ -41,35 +53,42 @@ defmodule Ecto.SoftDelete.Repo do
...
       end
 
-      def soft_delete(struct_or_changeset) do
-        struct_or_changeset
-        |> Ecto.Changeset.change(deleted_at: DateTime.utc_now())
-        |> update()
+      # Define soft_delete and soft_delete! with options
+      def soft_delete(struct_or_changeset, opts \\ []) do
+        changeset = Ecto.Changeset.change(struct_or_changeset, deleted_at: DateTime.utc_now())
+        __MODULE__.update(changeset, opts)
```
Comment (stuartjohnpage) [inline lib/ecto/soft_delete_repo.ex]: If you really wanted, you could probably do something like:

```suggestion
      def soft_delete(struct_or_changeset, opts \\ []) do
        struct_or_changeset
        |> Ecto.Changeset.change(deleted_at: DateTime.utc_now())
        |> __MODULE__.update(opts)
```
and keep those nice pipes!

Comment (danicodes01) [reply lib/ecto/soft_delete_repo.ex]: You are right! There is no need to assign that intermediate variable. Using a pipeline is a way cleaner! Also I should stick with the style of the rest of the repo. Making this change now. Thanks for all your input and help on this!

Comment (stuartjohnpage) [reply lib/ecto/soft_delete_repo.ex]: no worries, my pleasure!
